File: ruby/ext/create.c
#include "copyright.h"

/* Commands that create new objects */

#include "db.h"
#include "config.h"
#include "interface.h"
#include "stringutil.h"
#include "predicates.h"
#include "create.h"
#include "match.h"

/* utility for open and link */
static dbref parse_linkable_room(dbref player, const char *room_name)
{
    dbref room;

    /* parse room */
    if(!string_compare(room_name, "here")) {
	room = db[player].location;
    } else if(!string_compare(room_name, "home")) {
	return HOME;		/* HOME is always linkable */
    } else {
	room = parse_dbref(room_name);
    }

    /* check room */
    if(room < 0 || room >= db_top
       || Typeof(room) != TYPE_ROOM) {
	notify(player, "That's not a room!");
	return NOTHING;
    } else if(!can_link_to(player, room)) {
	notify(player, "You can't link to that.");
	return NOTHING;
    } else {
	return room;
    }
}

/* use this to create an exit */
void do_open(dbref player, const char *direction, const char *linkto)
{
    dbref loc;
    dbref exit;

#ifdef RESTRICTED_BUILDING
    if(!Builder(player)) {
	notify(player, "That command is restricted to authorized builders.");
	return;
    }
#endif /* RESTRICTED_BUILDING */

    if((loc = getloc(player)) == NOTHING) return;

    if(!*direction) {
		notify(player, "Open where?");
		return;
    } else if(!ok_name(direction)) {
		notify(player, "That's a strange name for an exit!");
		return;
    }

    if(!controls(player, loc)) {
		notify(player, "Permission denied.");
    } else if(!payfor(player, EXIT_COST)) {
		notify(player,
	       "Sorry, you don't have enough pennies to open an exit.");
    } else {
		/* create the exit */
		exit = new_object();
	
		/* initialize everything */
		db[exit].name = alloc_string(direction);
		db[exit].owner = player;
		db[exit].flags = TYPE_EXIT;
	
		/* link it in */
		PUSH(exit, db[loc].exits);
	
		/* and we're done */
		notify(player, "Opened.");
	
		/* check second arg to see if we should do a link */
		if(*linkto != '\0') {
			notify(player, "Trying to link...");
			if((loc = parse_linkable_room(player, linkto)) != NOTHING) {
				if(!payfor(player, LINK_COST)) {
					notify(player, "You don't have enough pennies to link.");
				} else {
					/* it's ok, link it */
					db[exit].location = loc;
					notify(player, "Linked.");
				}
			}
		}
    }
}

/* use this to link to a room that you own */
/* it seizes ownership of the exit */
/* costs 1 penny */
/* plus a penny transferred to the exit owner if they aren't you */
/* you must own the linked-to room AND specify it by room number */
void do_link(dbref player, const char *name, const char *room_name)
{
    dbref loc;
    dbref thing;
    dbref room;

    if((loc = getloc(player)) == NOTHING) return;
    if((room = parse_linkable_room(player, room_name)) == NOTHING) return;

    init_match(player, name, TYPE_EXIT);
    match_exit();
    match_neighbor();
    match_possession();
    match_me();
    match_here();
    if(Wizard(player)) {
	match_absolute();
	match_player();
    }

    if((thing = noisy_match_result()) != NOTHING) {
	switch(Typeof(thing)) {
	  case TYPE_EXIT:
	    /* we're ok, check the usual stuff */
	    if(db[thing].location != NOTHING) {
			if(controls(player, thing)) {
				if(Typeof(db[thing].location) == TYPE_PLAYER) {
				notify(player, "That exit is being carried.");
				} else {
				notify(player, "That exit is already linked.");
				}
			} else {
				notify(player, "Permission denied.");
			}
	    } else {
			/* handle costs */
			if(db[thing].owner == player) {
				if(!payfor(player, LINK_COST)) {
					notify(player, "It costs a penny to link this exit.");
					return;
				}
			} else {
				if(!payfor(player, LINK_COST + EXIT_COST)) {
					notify(player, "It costs two pennies to link this exit.");
					return;
	#ifdef RESTRICTED_BUILDING
				} else if(!Builder(player)) {
					notify(player, "Only authorized builders may seize exits.");
	#endif /* RESTRICTED_BUILDING */			
				} else {
					/* pay the owner for his loss */
					db[db[thing].owner].pennies += EXIT_COST;
				}
			}
	
			/* link has been validated and paid for; do it */
			db[thing].owner = player;
			db[thing].location = room;
	
			/* notify the player */
			notify(player, "Linked.");
	    }
	    break;
	  case TYPE_THING:
	  case TYPE_PLAYER:
	    if(!controls(player, thing)) {
		notify(player, "Permission denied.");
	    } else if(room == HOME) {
		notify(player, "Can't set home to home.");
	    } else {
		/* do the link */
		db[thing].exits = room; /* home */
		notify(player, "Home set.");
	    }
	    break;
	  case TYPE_ROOM:
	    if(!controls(player, thing)) {
		notify(player, "Permission denied.");
	    } else {
		/* do the link, in location */
		db[thing].location = room; /* dropto */
		notify(player, "Dropto set.");
	    }
	    break;
	  default:
	    notify(player, "Internal error: weird object type.");
	    fprintf(stderr, "PANIC weird object: Typeof(%d) = %d\n",
		    thing, Typeof(thing));
	    break;
	}
    }
}

/* use this to create a room */
void do_dig(dbref player, const char *name)
{
    dbref room;
    char buf[BUFFER_LEN];

#ifdef RESTRICTED_BUILDING
    if(!Builder(player)) {
	notify(player, "That command is restricted to authorized builders.");
	return;
    }
#endif /* RESTRICTED_BUILDING */

    /* we don't need to know player's location!  hooray! */
    if(*name == '\0') {
	notify(player, "Dig what?");
    } else if(!ok_name(name)) {
	notify(player, "That's a silly name for a room!");
    } else if(!payfor(player, ROOM_COST)) {
	notify(player, "Sorry, you don't have enough pennies to dig a room.");
    } else {
	room = new_object();

	/* Initialize everything */
	db[room].name = alloc_string(name);
	db[room].owner = player;
	db[room].flags = TYPE_ROOM;

	sprintf(buf, "%s created with room number %d.", name, room);
	notify(player, buf);
    }
}

/* use this to create an object */
void do_create(dbref player, char *name, int cost)
{
    dbref loc;
    dbref thing;

#ifdef RESTRICTED_BUILDING
    if(!Builder(player)) {
	notify(player, "That command is restricted to authorized builders.");
	return;
    }
#endif /* RESTRICTED_BUILDING */

    if(*name == '\0') {
	notify(player, "Create what?");
	return;
    } else if(!ok_name(name)) {
	notify(player, "That's a silly name for a thing!");
	return;
    } else if(cost < 0) {
	notify(player, "You can't create an object for less than nothing!");
	return;
    } else if(cost < OBJECT_COST) {
	cost = OBJECT_COST;
    }

    if(!payfor(player, cost)) {
	notify(player, "Sorry, you don't have enough pennies.");
    } else {
	/* create the object */
	thing = new_object();

	/* initialize everything */
	db[thing].name = alloc_string(name);
	db[thing].location = player;
	db[thing].owner = player;
	db[thing].pennies = OBJECT_ENDOWMENT(cost);
	db[thing].flags = TYPE_THING;

	/* endow the object */
	if(db[thing].pennies > MAX_OBJECT_ENDOWMENT) {
	    db[thing].pennies = MAX_OBJECT_ENDOWMENT;
	}

	/* home is here (if we can link to it) or player's home */
	if((loc = db[player].location) != NOTHING
	   && can_link_to(player, loc)) {
	    db[thing].exits = loc;	/* home */
	} else {
	    db[thing].exits = db[player].exits;	/* home */
	}

	/* link it in */
	PUSH(thing, db[player].contents);

	/* and we're done */
	notify(player, "Created.");
    }
}

File: ruby/lib/create.rb
require_relative 'helpers'

module TinyMud
    class Create
        include Helpers
        
        def initialize(db, notifier)
            @db = db
            @notifier = notifier
            @pred = Predicates.new(db, notifier)
            @match = Match.new(db, notifier)
        end
        
        #Given a room name or special keyowrds "here" or "home", parse_linkable_room attempts to find the room
        #player is asking to link to, and return the index for the room it is associated with (or tell the player they can't).
        def parse_linkable_room(player, room_name)
            if(room_name.downcase == "here")
                room = @db[player].location
            elsif(room_name.downcase == "home")
                return HOME
            else
                room = @db.parse_dbref(room_name)
            end
            
            #Check room
            if((room < 0) || (room >= @db.length) || (!room?(room)))
                @notifier.do_notify(player, Phrasebook.lookup('not-a-room'))
                return NOTHING
            elsif(!@pred.can_link_to(player, room))
                @notifier.do_notify(player, Phrasebook.lookup('bad-link'))
                return NOTHING
            else
                return room
            end
        end
        
        #do_open opens an exit belonging to the player in the specified direction.
        #Player must be a builder or higher, must supply a valid exit, and must have enough pennies.
        def do_open(player,direction,linkto)
            loc = getloc(player)

            if(getloc(player) == NOTHING)
                return
            end

            if(direction == nil)
                @notifier.do_notify(player, Phrasebook.lookup('open-where'))
                return
            elsif(!@pred.ok_name(direction))
                @notifier.do_notify(player, Phrasebook.lookup('strange-exit-name'))
                return
            end
            
            if(!@pred.controls(player,loc))
                @notifier.do_notify(player, Phrasebook.lookup('no-permission'))
            elsif(!@pred.payfor(player, EXIT_COST))
                @notifier.do_notify(player, Phrasebook.lookup('sorry-poor-open'))
            else
                exit = @db.add_new_record()
                room = @db[exit]
                room.name = direction
                room.owner = player
                room.flags = TYPE_EXIT
                
                @db[exit].next = @db[loc].exits
                @db[loc].exits = exit
                
                @notifier.do_notify(player, Phrasebook.lookup('opened'))
                
                if(linkto != nil)
                    @notifier.do_notify(player, Phrasebook.lookup('trying-to-link'))
                    loc = parse_linkable_room(player, linkto)
                    if(loc != NOTHING)
                        if(!@pred.payfor(player, LINK_COST))
                            @notifier.do_notify(player, Phrasebook.lookup('too-poor-to-link'))
                        else
                            #  Link the room.
                            @db[exit].location = loc
                            @notifier.do_notify(player, Phrasebook.lookup('linked'))
                        end
                    end
                end
            end
                
        end
        
        # Use this to link to a room that you own,
        # it seizes ownership of the exit.
        # costs 1 penny
        # plus a penny transferred to the exit owner if they aren't you.
        # you must own the linked-to room AND specify it by room number.
        def do_link(player, name, room_name)
            loc = getloc(player)
            return if (loc == NOTHING)

            room = parse_linkable_room(player, room_name)
            return if (room == NOTHING)

            @match.init_match(player, name, TYPE_EXIT)
            @match.match_exit()
            @match.match_neighbor()
            @match.match_possession()
            @match.match_me()
            @match.match_here()
            
            if (is_wizard(player)) 
                @match.match_absolute()
                @match.match_player()
            end
            
            thing = @match.noisy_match_result()     
            if (thing != NOTHING)
                case typeof(thing)
                    when TYPE_EXIT
                        #we're ok, check the usual stuff
                        if(@db[thing].location != NOTHING)
                            if(@pred.controls(player, thing))
                                if(player?(@db[thing].location))
                                    @notifier.do_notify(player, Phrasebook.lookup('exit-being-carried'))
                                else
                                    @notifier.do_notify(player, Phrasebook.lookup('exit-already-linked'))
                                end
                            else
                                @notifier.do_notify(player, Phrasebook.lookup('no-permission'))
                            end
                        else
                            if(@db[thing].owner == player)
                                if(!@pred.payfor(player, LINK_COST))
                                    @notifier.do_notify(player, Phrasebook.lookup('cost-penny-exit'))
                                    return
                                end
                            else
                                if(!@pred.payfor(player, LINK_COST + EXIT_COST))
                                    @notifier.do_notify(player, Phrasebook.lookup('cost-two-exit'))
                                    return
                                else
                                    #pay the owner for his loss
                                    @db[@db[thing].owner].pennies += EXIT_COST
                                end
                            end
                        
                            #link has been validated and paid for do it
                            @db[thing].owner = player
                            @db[thing].location = room
                            
                            #notify the player 
                            @notifier.do_notify(player, Phrasebook.lookup('linked'))
                        end
                    when TYPE_THING
                        if(!@pred.controls(player,thing))
                            @notifier.do_notify(player, Phrasebook.lookup('no-permission'))
                        elsif(room == HOME)
                            @notifier.do_notify(player, Phrasebook.lookup('no-set-home'))
                        else
                            #Activate link
                            @db[thing].exits = room 
                            @notifier.do_notify(player, Phrasebook.lookup('home-set'))
                        end
                    when TYPE_PLAYER # todo: no drop-through in ruby, this is a copy of the above
                        if(!@pred.controls(player,thing))
                            @notifier.do_notify(player, Phrasebook.lookup('no-permission'))
                        elsif(room == HOME)
                            @notifier.do_notify(player, Phrasebook.lookup('no-set-home'))
                        else
                            #Activate link
                            @db[thing].exits = room 
                            @notifier.do_notify(player, Phrasebook.lookup('home-set'))
                        end
                    when TYPE_ROOM
                        if(!@pred.controls(player,thing))
                            @notifier.do_notify(player, Phrasebook.lookup('no-permission'))
                        else
                            @db[thing].location = room
                            @notifier.do_notify(player, Phrasebook.lookup('drop-to-set'))
                        end
                    else #None of the types.    
                        @notifier.do_notify(player, Phrasebook.lookup('internal-error'))
                        $stderr.puts("PANIC weird object: typeof(thing) = #{typeof(thing)}\n")
                end
            end
        end
        
        #do_create creates an object with a particular name under the ownership of a player.
        #Creating an object costs penies.
        def do_create(player,name,cost)
            if(name == nil)
                @notifier.do_notify(player, Phrasebook.lookup('create-what'))
                return
            elsif(!@pred.ok_name(name))
                @notifier.do_notify(player, Phrasebook.lookup('silly-thing-name'))
                return
            elsif(cost < 0)
                @notifier.do_notify(player, Phrasebook.lookup('objects-must-have-a-value'))
                return
            elsif(cost < OBJECT_COST)
                cost = OBJECT_COST
            end
            
            
            
            if (!@pred.payfor(player, cost))
                @notifier.do_notify(player, Phrasebook.lookup('sorry-poor'))
            else
                #Okay, create the object and initialize it.
                thing = @db.add_new_record()
                thing_record = @db[thing]
                
                thing_record.name = name
                thing_record.location = player
                thing_record.owner = player
                thing_record.pennies = endow(cost)
                thing_record.flags = TYPE_THING
                
                #Make sure endowment isn't higher than max.
                if(thing_record.pennies > MAX_OBJECT_ENDOWMENT)
                    thing_record.pennies = MAX_OBJECT_ENDOWMENT
                end
                
                
                player_record = @db[player]
                if (player_record.location != NOTHING && @pred.can_link_to(player, player_record.location))
                    thing_record.exits = player_record.location
                else
                    thing_record.exits = player_record.exits
                end
                
                #PUSH(thing, db[player].contents)
                #define PUSH(thing, locative) \ ((db[(thing)].next = (locative)), (locative) = (thing))
                thing_record.next = player_record.contents
                player_record.contents = thing
                
                @notifier.do_notify(player, Phrasebook.lookup('created'))
            end
                
        end
        
        
        #Endow is a helper function to calculate the automatic endowment for an object. Originally in config.h.
        def endow(cost)
            return (cost - ENDOWMENT_CALCULATOR)/ENDOWMENT_CALCULATOR
        end
        
        
        #do_dig digs into an area, creating a new room.
        def do_dig(player,name)
            if(name == nil)
                @notifier.do_notify(player, Phrasebook.lookup('dig-what'))
            elsif(!@pred.ok_name(name))
                @notifier.do_notify(player, Phrasebook.lookup('silly-room-name'))
            elsif(!@pred.payfor(player, ROOM_COST))
                @notifier.do_notify(player, Phrasebook.lookup('sorry-poor-dig'))
            else
                #Everything is okay, create and initialize room
                room = @db.add_new_record()
                room_record = @db[room]
                
                
                room_record.name = name
                room_record.owner = player
                room_record.flags = TYPE_ROOM
                
                @notifier.do_notify(player, Phrasebook.lookup('created-room', name, room))
            end
        end
    end
end
